swarm/api/http/server.go

```diff
@@ -290,6 +291,117 @@ func (s *Server) HandleDelete(w http.ResponseWriter, r *Request) {
 	fmt.Fprint(w, newKey)
 }
 
+func (s *Server) HandlePostResource(w http.ResponseWriter, r *Request) {
+	var outdata string
+	if r.uri.Path != "" {
+		frequency, err := strconv.ParseUint(r.uri.Path, 10, 64)
+		if err != nil {
+			s.BadRequest(w, r, fmt.Sprintf("Cannot parse frequency parameter: %v", err))
+			return
+		}
+		err = s.api.ResourceCreate(r.uri.Addr, frequency)
+		if err != nil {
+			s.Error(w, r, fmt.Errorf("Resource creation failed: %v", err))
+			return
+		}
+		manifestKey, err := s.api.NewManifest()
+		if err != nil {
+			s.Error(w, r, fmt.Errorf("create manifest err: %v", err))
+			return
+		}
+		newKey, err := s.updateManifest(manifestKey, func(mw *api.ManifestWriter) error {
+			key, err := mw.AddEntry(bytes.NewReader([]byte(r.uri.Addr)), &api.ManifestEntry{
+				Path:        r.uri.Addr,
+				ContentType: api.ResourceContentType,
+				Mode:        0644,
+				Size:        int64(len(r.uri.Addr)),
+				ModTime:     time.Now(),
+			})
+			if err != nil {
+				return err
+			}
+			s.logDebug("resource manifest for for %s stored", key.Log())
+			return nil
+		})
+		if err != nil {
+			s.Error(w, r, fmt.Errorf("update manifest err: %v", err))
+			return
+		}
+		log.Debug("manifests", "new", newKey, "old", manifestKey)
+		outdata = fmt.Sprintf("%s", newKey)
+	}
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		s.Error(w, r, err)
+		return
+	}
+	_, _, _, err = s.api.ResourceUpdate(r.uri.Addr, data)
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
```

Comment:
I think `ResourceUpdate` should return a specific error to trigger a 401 response, otherwise this will just hide other errors.